### ISSUE TITLE
DWYU Doesn't take into account `includes` dirs listed in cc_* rules when resolving headers

### ISSUE THREAD

**Issue by mjj47** [martis42/depend_on_what_you_use#80]
Super excited to stumble onto your project :) I wrote a very similar tool for a hackathon! Happy to see your version integrated as a bazel aspect vs a tool built exclusively on top of bazel queries :P 

https://github.com/martis42/depend_on_what_you_use/blob/main/src/analyze_includes/evaluate_includes.py#L98
I believe this line only allows for resolving headers that are full paths to the root of a workspace, however, if I have 

```
- a/BUILD
cc_binary(
   name = "a",
   deps = "//b:headers"
)
- a/foo.cc
#include <b.h>
- b/BUILD
cc_library(
   name = "headers",
   hdrs = ["include/b.h"],
   includes = ["include"],
)
- b/include/b.h
```
This fails to resolve //a:a as using b.h from //b:b

**Comment by mjj47** [comment]
Ah, this could be my companies fault. I see you are resolving the different include paths here:
https://github.com/martis42/depend_on_what_you_use/blob/main/src/aspect/dwyu.bzl#L37

I added the print statement on line 47
```    print("Getting all includes for: {}, system_includes {},  header: {}".format(label, system_includes, header_file.path))```

And, for the case outlined above, I am seeing:
`Getting all includes for: //b:b, system_includes depset([".", "bazel-out/k8-fastbuild/bin", "bazel-out/k8-fastbuild/b/include",]), header:  b/include/b.h`

Looks like the aspect attempts to match based on path in tree, while the system includes are path-in-bazel-output.

**Comment by mjj47** [comment]
Didn't have permissions to push a new branch for a PR - thoughts on the following patch? 

[0001-Add-dep-CcInfo-.compilation_context.includes-to-list.patch](https://github.com/martis42/depend_on_what_you_use/files/10387796/0001-Add-dep-CcInfo-.compilation_context.includes-to-list.patch)

**Comment by storypku** [comment]
> ve //a:a

Hi @mjj47 , DWYU gives correct behavior on your testcase.

See below

a/BUILD

```
load("@rules_cc//cc:defs.bzl", "cc_binary")

package(default_visibility = ["//visibility:public"])

cc_binary(
    name = "a",
    srcs = ["a.cc"],
    deps = [
        "//b:b",
    ],
)
```

a/a.cc
```
#include <iostream>
#include "b.h"

int main() {
    std::cout << "The answer is: " << kAnswer << std::endl;
    return 0;
}
```

b/BUILD 
```
package(default_visibility = ["//visibility:public"])

cc_library(
        name = "b",
        hdrs = ["include/b.h"],
        includes = ["include"],
)
```

b/include/b.h

```
#pragma once

inline constexpr int kAnswer = 42;
```
Running dwyu on //a:a:

```
$ bazel-5.3.2 --bazelrc=/dev/null build --aspects=//test/aspect:aspect.bzl%dwyu_default_aspect --output_groups=cc_dwyu_output //a:a
DEBUG: /work/github/depend_on_what_you_use.git/src/aspect/dwyu.bzl:46:10: Getting all includes for: //b:b, system_includes depset(["b/include", "bazel-out/k8-fastbuild/bin/b/include"]),  header: b/include/b.h
INFO: Analyzed target //a:a (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Aspect //test/aspect:aspect.bzl%dwyu_default_aspect of //a:a up-to-date:
  bazel-bin/a/a_a_dwyu_report.json
INFO: Elapsed time: 0.056s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
```

**Comment by storypku** [comment]
> Didn't have permissions to push a new branch for a PR - thoughts on the following patch?
> 
> [0001-Add-dep-CcInfo-.compilation_context.includes-to-list.patch](https://github.com/martis42/depend_on_what_you_use/files/10387796/0001-Add-dep-CcInfo-.compilation_context.includes-to-list.patch)

Like @mjj47 , I have puzzles in this regard. Shall we also consider `system_includes`, `includes`, `quote_includes`, and even `framework_includes` ? Ref: https://github.com/grailbio/bazel-compilation-database/blob/940cedacdb8a1acbce42093bf67f3a5ca8b265f7/aspects.bzl

**Comment by martis42** [comment]
@mjj47 
Thanks for the kind words. I am happy you like the tool and it is helpful to you.

I can't reproduce your problem. One of the integration tests covers this scenario: https://github.com/martis42/depend_on_what_you_use/blob/main/test/aspect/includes/BUILD
The test framework might seem a bit obscure to someone not familiar with it. Essentially I test that the following command returns without error:
`bazel build --aspects=//test/aspect:aspect.bzl%dwyu_default_aspect --output_groups=cc_dwyu_output //test/aspect/includes:use_includes`

In case I understand your problem wrongly, please correct me.

> Didn't have permissions to push a new branch for a PR - thoughts on the following patch?

This project follows the default GitHub way (at least to my understanding) of requiring forks for contribution.
More details are described [here](https://docs.github.com/en/get-started/quickstart/contributing-to-projects)

**Comment by martis42** [comment]
@storypku 

> Like @mjj47 , I have confusion in this regard. Shall we also consider system_includes, includes, quote_includes, and even framework_includes ? Ref: https://github.com/grailbio/bazel-compilation-database/blob/940cedacdb8a1acbce42093bf67f3a5ca8b265f7/aspects.bzl

True, there are more values available from `CompilationContext` than DWYU is currently using. Then again, all test cases I could think of work with the current implementation including various combinations of the cc_library options for modifying includes paths, external workspaces and generated code.
I would not change the implementation unless there is a use case which requires this.

From what I remember during the initial DWYU development:
- `includes` is relevant for the cc_library attributes include_prefix and strip_include_prefix. Those use cases are supported, but DWYU exploits that those virtual include paths are also part of the header paths in the header lists
- `system_includes` is relevant for the cc_library includes and is utilized by DWYU
- `quote_includes` is useless as far as I can tell. This points always to the same places for all scenarios I could think of.
- `framework_includes` I have no idea what this is about :man_shrugging: 

`CompilationContext` has so many values since there are many nuances in telling the compiler how to use which header exactly. However, DWYU works on a simpler level. We "only" have to know at which path a header is available at all. If it is included with brackets or quotes or if compiler warnings should be reported or not does not matter to DWYU.

**Comment by mjj47** [comment]
Thanks for the super helpful responses! After hearing back, I assumed this was my custom toolchain's fault, but, I was able to run your nested test cases with no issue - even made a new test case in tree of DWYU that mirrored my exact failure mode, however, that test case passed too... I then thought it could be related to using a cc_test or cc_binary rule rather than a cc_library, but still, couldn't repro the failure. 

I would *think* for the sake of correctness, including `includes` in your aspect would be valid, but I understand not wanting to adopt it without a provable failure mode. 

I will definitely keep looking for the root cause of my case, but, I would understand closing this as won't fix until I can come with a test that fails within DWYU source :) 

(And good to know on the permissions - I think its 100% a problem on my end. My company has a private github mirror, I bet I have some tokens configured for that endpoint rather than public github.com. Definitely will keep trying to push up PRs because I am highly interested in adding support for alias's, adding to the fixer tool to auto add deps instead of just remove, enable running the aspect and fixer against a full package at a time vs single target at a time)

**Comment by martis42** [comment]
@mjj47 
I am always happy about contributions 😁

Regarding the concrete points you mention. Please note I am already working on automatically adding missing dependencies with the automatic fixes tool. Thus, you should not invest time into that in parallel.

> enable running the aspect and fixer against a full package at a time vs single target at a time)

I don't understand your exact problem. The aspect is able to run on full packages. Simply use `//foo:all` as target pattern. Or `//foo/...` for foo and all packages below it.

You can also lint a target and all it's transitive deps by using the Recursion feature described in the Readme.

The fixing tool always works on a whole workspace. Simply looking for whatever DWYU report files are in the Bazel output dir available.
